File: go/problem_493.go
package main

import (
	"fmt"
	"math/big"
	"sort"
)

const (
	numColors    = 7
	numPerColor  = 10
	numToPick    = 20
	sumAllPicked = numColors*numPerColor - numToPick
)

type counts struct {
	colors, picks *big.Int
}

var cache = make(map[[numColors]int]counts)

func allBallsPicked(urn *[numColors]int) bool {
	sum := 0
	for _, v := range urn {
		sum += v
	}
	return sum == sumAllPicked
}

func countColors(urn *[numColors]int) int {
	count := 0
	for _, v := range urn {
		if v < numPerColor {
			count++
		}
	}
	return count
}

func computeCounts(urn *[numColors]int) counts {
	var key [numColors]int = *urn
	sort.Ints(key[:])
	c, ok := cache[key]
	if ok {
		return c
	}

	if allBallsPicked(urn) {
		c = counts{big.NewInt(int64(countColors(urn))), big.NewInt(1)}
	} else {
		c = counts{big.NewInt(0), big.NewInt(0)}
		for i := 0; i < numColors; i++ {
			for j := 0; j < urn[i]; j++ {
				urn[i]--
				returned := computeCounts(urn)
				c.colors.Add(c.colors, returned.colors)
				c.picks.Add(c.picks, returned.picks)
				urn[i]++
			}
		}
	}
	cache[key] = c
	return c
}

func main() {
	var urn [numColors]int
	for i := 0; i < numColors; i++ {
		urn[i] = numPerColor
	}
	counts := computeCounts(&urn)
	fmt.Printf("Total colors = %s\n", counts.colors.String())
	fmt.Printf("Total picks = %s\n", counts.picks.String())
	fmt.Printf("Cache size = %d\n", len(cache))
	average := new(big.Float).Quo(
		new(big.Float).SetInt(counts.colors),
		new(big.Float).SetInt(counts.picks))
	fmt.Printf("Average = %0.9f\n", average)
}

File: py/problem_493.py
#!/usr/bin/env python 

from functools import cache

NUM_COLORS = 7
NUM_PER_COLOR = 10
NUM_TO_PICK = 20
SUM_ALL_PICKED = NUM_COLORS * NUM_PER_COLOR - NUM_TO_PICK


def all_balls_picked(urn):
  return sum(urn) == SUM_ALL_PICKED
    

def count_colors(urn):
  return sum(val < NUM_PER_COLOR for val in urn)


def pick(urn, color):
  return [count - 1 if index == color else count 
          for index, count in enumerate(urn) ]


def sum_tuples(iters):
  return tuple(map(sum, zip(*iters)))
  

def compute_counts(urn):
  return compute_counts_cached(tuple(sorted(urn)))


@cache
def compute_counts_cached(urn):
  if all_balls_picked(urn):
    return (count_colors(urn), 1)
  else:
    return sum_tuples(compute_counts(pick(urn, i))
                      for i in range(len(urn))
                      for _ in range(urn[i]))

  
if __name__ == '__main__':
  starting_urn = [NUM_PER_COLOR] * NUM_COLORS
  counts = compute_counts(starting_urn)
  print("Total colors = {}".format(counts[0]))
  print("Total picks = {}".format(counts[1]))
  print("Average = {0:0.9f}".format(float(counts[0]) / counts[1]))
